### ISSUE TITLE
RF/GBDT Leaf-wise Tree Growth Instead of Level-Wise Tree Growth

### ISSUE THREAD

**Issue by zhangpengshan** [ShifuML/shifu#249]
https://github.com/Microsoft/LightGBM/wiki/Features

Pros: Leaf-wise algorithm can reduce more loss than level-wise algorithm.

Each iteration in master, todo queue is sorted by delta loss and then send todo nodes to Worker.

Instead of max-depth, set max-leaves as an parameter to stop tree growth.

**Comment by zhangpengshan** [comment]
Queue can be sorted by improvement, which can be found in below scikit-learn:
weighted_rate * gain = improvement
weighted_rate = weighted_node_cnt / weighted_total_cnt

https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/tree/_criterion.pyx#L203
